core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/StageExecutionStats.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.scheduler.faulttolerant;
+
+import com.google.inject.Inject;
+import io.airlift.stats.CounterStat;
+import org.weakref.jmx.MBeanExporter;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class StageExecutionStats
+{
+    private final Map<String, CounterStat> outputEstimationKindCounters = new ConcurrentHashMap<>();
+    private final MBeanExporter mbeanExporter;
+
+    @Inject
+    public StageExecutionStats(MBeanExporter mbeanExporter)
+    {
+        this.mbeanExporter = mbeanExporter;
+    }
+
+    public void recordSourceOutputEstimationOnStageStart(String sourceOutputEstimationKind, int sourcesCount)
+    {
+        updateSourceOutputEstimationKindCounter(sourceOutputEstimationKind, sourcesCount);
+    }
+
+    public void recordSourcesFinishedOnStageStart(int sourcesCount)
+    {
+        updateSourceOutputEstimationKindCounter("finished", sourcesCount);
+    }
+
+    private void updateSourceOutputEstimationKindCounter(String outputEstimationKind, int sourcesCount)
+    {
+        getCounterStat(outputEstimationKind).update(sourcesCount);
+    }
+
+    private CounterStat getCounterStat(String outputEstimationKind)
+    {
+        return outputEstimationKindCounters.computeIfAbsent(outputEstimationKind, ignored -> {
+            CounterStat counter = new CounterStat();
+            mbeanExporter.exportWithGeneratedName(counter, StageExecutionStats.class, "output_size_estimation" + "_" + outputEstimationKind);
```

Comment:
will mbeanExporter unexport this later?

is this a problem in tests executing concurrently?

Reply:
> will mbeanExporter unexport this later?

Are we unexporting jmx ever? If we do I would expect it will unexport these at same time we unexport others. But it would be at app shutdown anyway - so does not matter I think.

> is this a problem in tests executing concurrently?

If we depended on stats in tests - probably yeah - it would cause issues. But we do not.

Reply:
how many different `outputEstimationKind` values there are? <10, right?

Reply:
Yeah a couple right now. Don't expect the number to grow significantly